volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -213,65 +213,63 @@ def get_elf_table_name(self):
         )
         return elf_table_name
 
-    def get_symbols(self):
+    def get_symbols(
+        self,
+    ) -> Iterable[Tuple[int, interfaces.objects.ObjectInterface]]:
         """Get symbols of the module
 
         Yields:
-                A symbol object
+            A tuple containing the ELF symbol index and the corresponding ELF symbol object
         """
 
-        if not hasattr(self, "_elf_table_name"):
-            self._elf_table_name = self.get_elf_table_name()
-        if symbols.symbol_table_is_64bit(self._context, self.get_symbol_table_name()):
-            prefix = "Elf64_"
-        else:
-            prefix = "Elf32_"
-        syms = self._context.object(
-            self.get_symbol_table_name() + constants.BANG + "array",
+        if not self.section_strtab or self.num_symtab < 1:
+            return None
+
+        elf_table_name = self.get_elf_table_name()
+        symbol_table_name = self.get_symbol_table_name()
+
+        is_64bit = symbols.symbol_table_is_64bit(self._context, symbol_table_name)
+        sym_name = "Elf64_Sym" if is_64bit else "Elf32_Sym"
+        sym_type = self._context.symbol_space.get_type(
+            elf_table_name + constants.BANG + sym_name
+        )
+        elf_syms = self._context.object(
+            symbol_table_name + constants.BANG + "array",
             layer_name=self.vol.layer_name,
             offset=self.section_symtab,
-            subtype=self._context.symbol_space.get_type(
-                self._elf_table_name + constants.BANG + prefix + "Sym"
-            ),
-            count=self.num_symtab + 1,
+            subtype=sym_type,
+            count=self.num_symtab,
         )
-        if self.section_strtab:
-            yield from syms
+        for elf_sym_num, elf_sym_obj in enumerate(elf_syms):
+            # Prepare the symbol object for methods like get_name()
+            elf_sym_obj.cached_strtab = self.section_strtab
```

Comment:
I'm not a massive fan of the `cached_strtab` attribute, since the implication is that it's set on creation, but it isn't part of the constructor (meaning an `elf_sym` could be constructed but not have it set) but this isn't the place to resolve that.   I see there's a similar design pattern for `elf_phdr` so probably one we want to figure out some time but the method used is the only one I've got off the top of my head...

Reply:
Yeah, I'm not a fan of that either, but that's how it was implemented :). I try not to be too tempted to rewrite everything I don't like, hehe. Instead, I focus on reusing and keeping things as they are while working towards the main goal of the PR. Although, many times I fail and end up refactoring more than I should... I know.. and I guess you know that by now, haha .. sorry about that. But this time I didn't go down that route. Happy we can explore a better implementation for that in a future PR.